File: Room.py
class Room:
    #ROOM class for thermodynamic simulation

    def __init__(ID,T_range,L,W,building):
        #defining constants
        self.rho_air = 1.23 #kg m^-3
        self.cp_air = 1004 #J kg^-1 K^-1 
        self.cv_air = 717 #J kg^-1 K^-1
        self.a = 0.9 #absorptivity of room
        self.e = 0.9 #emissivity of room 
        self.sb = 5.67 * 10^-8 #stefan boltzmann constant

        self.ID = ID
        self.T_range = T_range
        self.L = L
        self.W = W
        self.building = building
        self.heater = building.heater
        self.cooler = building.cooler

#how to fit in these as well? do you have to say = undefined?
        self.T # Current temperature
        self.H=3 # Height (m)
        self.walls = [] # List of structs: [otherside, area, R_eff]
        self.floor # struct: [ground, area, R]
        self.roof # struct: [outside, area, R]

    @property
    def V(self):
        # Computes volume of the room
        V = self.L*self.W*self.H
    @property
    def SA(self):
        # Computes surface area of the room
        SA = sum([self.walls.area]) + self.floor.area + self.roof.area
        
    def dTdt(self,t,T): # Computers overall dTdt for the room
        dQdt_cc = self.getCC(t,T) # Gets conductive/convective heat transfer amt
        dQdt_LW_rad = self.getLWRadiation(t,T) # gets LW radiative heat transfer amount
        dQdt_SW_rad = self.getSWRadiation(t) # gets SW radiative heat transfer amount
        dQdt_internal = self.getInternal(t) # gets internal heat generation rate
        dQdt_heater = self.getHeating(t,T) #gets the heating amount for this room
        dQdt_cooler = self.getCooling(t,T) #gets the cooling amount for this room
        dTdt = (24*3600/(self.rho_air * self.cv_air * self.V)) * \
            (dQdt_cc + dQdt_LW_rad + dQdt_SW_rad + dQdt_internal \
            + dQdt_heater + dQdt_cooler)
        return dTdt

    def getCC(self,t,T):
        dQdt_cc = 0 
        roomTemp = T(self.ID)
        for wallidx in range(0,length(self.walls)):
            wall = self.walls(wallidx)
            if isinstance(wall.otherside,'Outside'): #need to replace isinstance function with pythonic fxn
                outsideT = wall.otherside.T(t)
                dQdt_cc = dQdt_cc + (wall.area * (outsideT - roomTemp)/wall.R)
            elif isinstance(wall.otherside,'Room'):
                otherRoomT = T(wall.otherside.ID)
                dQdt_cc = dQdt_cc + (wall.area * (otherRoomT - roomTemp)/wall.R)
            
        
        dQdt_cc = dQdt_cc + self.roof.area*(self.roof.outside.T(t) - roomTemp)/self.roof.R
        dQdt_cc = dQdt_cc + self.floor.area*(self.floor.ground.T(t) - roomTemp)/self.floor.R
        return dQdt_cc

    def getInternal(self,t):
        if self.ID == 7:
            dQdt_internal = 5000
        elif self.ID == 3 and isWorkHours(t):
            dQdt_internal = 2000
        else:
            dQdt_internal = 0
        return dQdt_internal
        
    
    def getLWRadiation(self,t,T):
        dQdt_LW_rad = 0
        roomTemp = T(self.ID)
        for wallidx in range(0,length(self.walls)):
            wall = self.walls(wallidx)
            if isinstance(wall.otherside,'Outside'):
                outside = wall.otherside
                dQdt_LW_rad = dQdt_LW_rad + wall.area*self.sb*(-self.e*(roomTemp**4) \
                    + self.a*(outside.T_sky(t)**4))
            elif isinstance(wall.otherside,'Room'):
                otherRoom = wall.otherside
                otherRoomT = T(otherRoom.ID)
                dQdt_LW_rad = dQdt_LW_rad + wall.area*self.sb*(-self.e*(roomTemp**4) \
                    + self.a*(otherRoomT**4))
        
        dQdt_LW_rad = dQdt_LW_rad +self.roof.area*self.sb*(-self.e*(roomTemp**4)\
            + self.a*(self.roof.outside.T_sky(t)**4))
        dQdt_LW_rad = dQdt_LW_rad +self.floor.area*self.sb*(-self.e*(roomTemp**4)\
            + self.a*(self.floor.ground.T(t)**4))
        return dQdt_LW_rad
    
    def getSWRadiation(self,t):
        dQdt_SW_rad = self.roof.area*self.a*self.roof.outside.S(t)
        return dQdt_SW_rad
    
    def addWall(self,otherside,A_wall,R_eff):
        newWall = {'otherside': otherside,'area':A_wall,'R':R_eff}
        self.walls = self.walls.append(newWall)
    
    def addFloor(self,ground,A_floor,R_eff):
        self.floor = {'ground':ground,'area':A_floor,'R':R_eff}
    
    def addRoof(self,outside,A_roof,R_eff):
        self.roof = {'outside':outside,'area':A_roof,'R':R_eff}
    
    def getHeating(self,t,T):
        # Gets the heater's output for time t, converts it
        # into units of energy/time
        roomTemp = T(self.ID)
        [TH,fH] = self.heater.getHeating(t,T)
        deltaT = TH - roomTemp
        roomFlow = fH(self.ID)
        dQdt_heater = self.cp_air * self.rho_air * roomFlow * deltaT
        return dQdt_heater
    
    def getCooling(self,t,T):
        roomTemp = T(self.ID)
        [TC,fC] = self.cooler.getCooling(t,T)
        deltaT = TC - roomTemp
        roomFlow = fC(self.ID)
        dQdt_cooler = self.cp_air * self.rho_air * roomFlow * deltaT
        return dQdt_cooler
    

    
    

File: Room.m
classdef Room < handle
    %ROOM class for thermodynamic simulation
    properties
        ID; % Numeric ID of the room
        T; % Current temperature
        T_range; % 'Acceptable' temperature range
        L; % Length (m)
        W; % Width (m)
        H=3; % Height (m)
        building; % the building object that this room belongs to
        walls; % List of structs: [otherside, area, R_eff]
        floor; % struct: [ground, area, R]
        roof; % struct: [outside, area, R]
        heater; % the heater object for the building
        cooler; % the AC object for the building
    end
    properties (Dependent)
        V; % Volume of the room
        SA; % Surface area of the room
    end
    properties (Constant)
        rho_air = 1.23; %kg m^-3
        cp_air = 1004; %J kg^-1 K^-1 
        cv_air = 717; %J kg^-1 K^-1
        a = 0.9; %absorptivity of room
        e = 0.9; %emissivity of room 
        sb = 5.67 * 10^-8; %stefan boltzmann constant
    end
    methods
        function r = Room(ID,T_range,L,W,building)
            r.ID = ID;
            r.T_range = T_range;
            r.L = L;
            r.W = W;
            r.building = building;
            r.heater = building.heater;
            r.cooler = building.cooler;
        end
        function dTdt = dTdt(obj,t,T) % Computers overall dTdt for the room
            dQdt_cc = obj.getCC(t,T); % Gets conductive/convective heat transfer amt
            dQdt_LW_rad = obj.getLWRadiation(t,T); % gets LW radiative heat transfer amount
            dQdt_SW_rad = obj.getSWRadiation(t); % gets SW radiative heat transfer amount
            dQdt_internal = obj.getInternal(t); % gets internal heat generation rate
            dQdt_heater = obj.getHeating(t,T); %gets the heating amount for this room
            dQdt_cooler = obj.getCooling(t,T); %gets the cooling amount for this room
            dTdt = (24*3600/(obj.rho_air * obj.cv_air * obj.V)) * ...
                (dQdt_cc + dQdt_LW_rad + dQdt_SW_rad + dQdt_internal ...
                + dQdt_heater + dQdt_cooler);
        end
        function dQdt_cc = getCC(obj,t,T)
            dQdt_cc = 0; 
            roomTemp = T(obj.ID);
            for wallidx = 1:length(obj.walls)
                wall = obj.walls(wallidx);
                if isa(wall.otherside,'Outside')
                    outsideT = wall.otherside.T(t);
                    dQdt_cc = dQdt_cc + (wall.area * (outsideT - roomTemp)/wall.R);
                elseif isa(wall.otherside,'Room')
                    otherRoomT = T(wall.otherside.ID);
                    dQdt_cc = dQdt_cc + (wall.area * (otherRoomT - roomTemp)/wall.R);
                end
            end
            dQdt_cc = dQdt_cc + obj.roof.area*(obj.roof.outside.T(t) - roomTemp)/obj.roof.R;
            dQdt_cc = dQdt_cc + obj.floor.area*(obj.floor.ground.T(t) - roomTemp)/obj.floor.R;
        end
        function dQdt_internal = getInternal(obj,t)
            if obj.ID == 7
                dQdt_internal = 5000;
            elseif (obj.ID == 3) && isWorkHours(t)
                dQdt_internal = 2000;
            else
                dQdt_internal = 0;
            end
        end
        function dQdt_LW_rad = getLWRadiation(obj,t,T)
            dQdt_LW_rad = 0;
            roomTemp = T(obj.ID);
            for wallidx = 1:length(obj.walls)
                wall = obj.walls(wallidx);
                if isa(wall.otherside,'Outside')
                    outside = wall.otherside;
                    dQdt_LW_rad = dQdt_LW_rad + wall.area*obj.sb*(-obj.e*(roomTemp^4) ...
                        + obj.a*(outside.T_sky(t)^4));
                elseif isa(wall.otherside,'Room')
                    otherRoom = wall.otherside;
                    otherRoomT = T(otherRoom.ID);
                    dQdt_LW_rad = dQdt_LW_rad + wall.area*obj.sb*(-obj.e*(roomTemp^4) ...
                        + obj.a*(otherRoomT^4));
                end
            end
            dQdt_LW_rad = dQdt_LW_rad +obj.roof.area*obj.sb*(-obj.e*(roomTemp^4)...
                + obj.a*(obj.roof.outside.T_sky(t)^4));
            dQdt_LW_rad = dQdt_LW_rad +obj.floor.area*obj.sb*(-obj.e*(roomTemp^4)...
                + obj.a*(obj.floor.ground.T(t)^4));
        end
        function dQdt_SW_rad = getSWRadiation(obj,t)
            dQdt_SW_rad = obj.roof.area*obj.a*obj.roof.outside.S(t);
        end
        function addWall(obj,otherside,A_wall,R_eff)
            newWall = struct('otherside',otherside,'area',A_wall,'R',R_eff);
            obj.walls = cat(1,obj.walls,newWall);
        end
        function addFloor(obj,ground,A_floor,R_eff)
            obj.floor = struct('ground',ground,'area',A_floor,'R',R_eff);
        end
        function addRoof(obj,outside,A_roof,R_eff)
            obj.roof = struct('outside',outside,'area',A_roof,'R',R_eff);
        end
        function dQdt_heater = getHeating(obj,t,T)
            % Gets the heater's output for time t, converts it
            % into units of energy/time
            roomTemp = T(obj.ID);
            [TH,fH] = obj.heater.getHeating(t,T);
            deltaT = TH - roomTemp;
            roomFlow = fH(obj.ID);
            dQdt_heater = obj.cp_air * obj.rho_air * roomFlow * deltaT;
        end
        function dQdt_cooler = getCooling(obj,t,T)
            roomTemp = T(obj.ID);
            [TC,fC] = obj.cooler.getCooling(t,T);
            deltaT = TC - roomTemp;
            roomFlow = fC(obj.ID);
            dQdt_cooler = obj.cp_air * obj.rho_air * roomFlow * deltaT;
        end
        function V = get.V(obj)
            % Computes volume of the room
            V = obj.L*obj.W*obj.H;
        end
        function SA = get.SA(obj)
            % Computes surface area of the room
            SA = sum([obj.walls.area]) + obj.floor.area + obj.roof.area;
        end
        
    end
    
end

